src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala

```diff
@@ -340,21 +340,48 @@ trait ContextErrors {
 
       //typedSelect
       def NotAMemberError(sel: Tree, qual: Tree, name: Name) = {
-        def errMsg = {
+        import util.{ EditDistance, StringUtil }
+        def errMsg: String = {
           val owner            = qual.tpe.typeSymbol
           val target           = qual.tpe.widen
           def targetKindString = if (owner.isTypeParameterOrSkolem) "type parameter " else ""
           def nameString       = decodeWithKind(name, owner)
           /* Illuminating some common situations and errors a bit further. */
           def addendum         = {
-            val companion = {
+            val companionSymbol: Option[Symbol] = {
               if (name.isTermName && owner.isPackageClass) {
                 target.member(name.toTypeName) match {
-                  case NoSymbol => ""
-                  case sym      => "\nNote: %s exists, but it has no companion object.".format(sym)
+                  case NoSymbol => None
+                  case sym      => Some(sym)
                 }
               }
-              else ""
+              else None
+            }
+            val companion = companionSymbol match {
+              case Some(sym) => "\nnote: %s exists, but it has no companion object.".format(sym)
+              case _         => ""
+            }
+            // find out all the names available under target within 2 edit distances
+            lazy val alternatives: Vector[String] = {
+              val editThreshold = 2
+              val x = name.toString
+              if (x.size < 2) Vector()
+              else {
+                val names = target.nonPrivateMembers.toList.toVector.map(_.name.toString).distinct
```

Comment:
`nonPrivateMembers.iterator.filterNot(m => m.isConstructor).map(_.name.decode).filter(_.length > 2).distinct.toVector`

This does it in a single traversal. Using the decoded name to avoid the issue below:

```
scala> object C { def <= = 0 }
defined object C
warning: previously defined class C is not a companion to object C.
Companions must be defined together; you may wish to use :paste mode for this.

scala> C.lesseq
         ^
       error: value lesseq is not a member of object C
       did you mean $less$eq?
```

I think filtering short names also here makes sense. Doing it after decoding removes operators like `<=`.

I think it still includes too much, for example

```
scala> case class C(x: Int)
scala> typeOf[C].nonPrivateMembers.iterator.filterNot(m => m.isConstructor).map(_.name.decode).filter(_.length > 2).distinct.toVector
res32: scala.collection.immutable.Vector[String] = Vector(equals, toString, hashCode, canEqual, productIterator, productElement, productArity, productPrefix, copy$default$1, copy, $asInstanceOf, $isInstanceOf, synchronized, finalize, wait, notifyAll, notify, clone, getClass, asInstanceOf, isInstanceOf)
```
We should exclude `copy$default$1`, `$asInstanceOf`, `$isInstanceOf`. But filtering all of `m.isSynthetic` is too much

```
scala> typeOf[C].nonPrivateMembers.iterator.filterNot(m => m.isSynthetic || m.isConstructor).map(_.name.decode).filter(_.length > 2).distinct.toVector
res33: scala.collection.immutable.Vector[String] = Vector(synchronized, finalize, wait, notifyAll, notify, clone, getClass, asInstanceOf, isInstanceOf)
```
We lose `copy`, `hashCode`, ...

Maybe we can do it by character (`.contains("$")`)? Since we're using decoded names, `$` should only show up in names that are not supposed to be used by the user.

Reply:
You could also delay the `toList` until after the filters below